spices/SPICE-0011-const-object-members.adoc

```diff
@@ -0,0 +1,111 @@
+= Const Object Members
+
+* Proposal: link:./SPICE-0011-const-object-members.adoc[SPICE-0011]
+* Author: https://github.com/bioball[Dan Chao]
+* Status: Review
+* Implemented in: TBD
+* Category: Language
+
+== Introduction
+
+This SPICE proposes allowing object members to be `const`.
+
+== Motivation
+
+Today, the `const` modifier can only be applied to class properties.
+
+However, there are times where this rule is insufficient.
+
+When a module amends another module, it has special rules; top-level properties are considered _object properties_, and not _class properties_.
+
+This means that these properties cannot be declared `const`.
+
+[source,pkl]
+----
+amends "foo.pkl"
+
+const bar = 5 // Pkl Error: modifier `const` cannot be applied to object members
+----
+
+Amending modules can also contain classes and typealiases.
+These members must be declared with the `local` modifier.
+
+However, these members cannot reference their enclosing module's properties, because classes and typealiases can only reference properties that are `const`.
+
+One possible scenario is to define a test fixture.
+
+[source,pkl]
+----
+amends "pkl:test"
+
+import "birds.pkl"
+
+local warbler = "Warbler"
+
+local class Warbler extends birds.Bird {
+  name = warbler // Pkl Error: cannot reference `warbler` from here because it is not `const`
+}
+----
+
+There is no straightforward way to resolve this, because `warbler` cannot be declared `const`:
+
+[source,pkl]
+----
+amends "pkl:test"
+
+import "birds.pkl"
+
+local const warbler = "Warbler" // Pkl Error: modifier `const` cannot be applied to object members
+
+local class Warbler extends birds.Bird {
+  name = warbler
+}
+----
+
+== Proposed Solution
+
+To fix this papercut, a object members are allowed to be `const`, but must co-appear with the `local` modifier.
+
+[source,pkl]
+----
+amends "foo.pkl"
+
+const bar = "bar" // Pkl Error: modifier `const` can only be applied to object members that are also `local`.
```

Comment:
As an illustration of the above; `bar` is only an object member if it is `local`.

Reply:
We actually treat any property defined on an amending module as an object property, rather than a class property.

See the two branches when we parse a module:

Branch 1:
https://github.com/apple/pkl/blob/4b4d81ba9380a1a47c5cab94e654858dafb64482/pkl-core/src/main/java/org/pkl/core/ast/builder/AstBuilder.java#L188-L213

Branch 2:
https://github.com/apple/pkl/blob/4b4d81ba9380a1a47c5cab94e654858dafb64482/pkl-core/src/main/java/org/pkl/core/ast/builder/AstBuilder.java#L215-L234

One parses as class properties, one parses as object properties.